src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+}
+
+/**
+ * A module that will enable near interaction pointer selection for hands and motion controllers of XR Input Sources
+ */
+export class WebXRNearInteraction extends WebXRAbstractFeature {
+    private static _idCounter = 200;
+
+    private _attachController = (xrController: WebXRInputSource) => {
+        if (this._controllers[xrController.uniqueId]) {
+            // already attached
+            return;
+        }
+        const { hoverIndexMeshTip, pickIndexMeshTip } = this._generateNewHandTipMeshes();
+
+        // get two new meshes
+        this._controllers[xrController.uniqueId] = {
+            xrController,
+            meshUnderPointer: null,
+            pick: null,
+            pickIndexMeshTip,
+            hoverIndexMeshTip,
+            grabRay: new Ray(new Vector3(), new Vector3()),
+            nearInteraction: false,
+            nearGrab: false,
+            nearGrabInProcess: false,
+            id: WebXRNearInteraction._idCounter++,
+        };
+
+        if (this._attachedController) {
+            if (
+                !this._options.enablePointerSelectionOnAllControllers &&
+                this._options.preferredHandedness &&
+                xrController.inputSource.handedness === this._options.preferredHandedness
+            ) {
+                this._attachedController = xrController.uniqueId;
+            }
+        } else {
+            if (!this._options.enablePointerSelectionOnAllControllers) {
+                this._attachedController = xrController.uniqueId;
+            }
+        }
+        // NOT SURE: if i should return null for gaze and screen modes
+        switch (xrController.inputSource.targetRayMode) {
+            case "tracked-pointer":
+                return this._attachTrackedPointerRayMode(xrController);
+            case "gaze":
+                return null;
+            case "screen":
+                return null;
+        }
+    };
+
+    private _controllers: {
+        [controllerUniqueId: string]: {
+            xrController?: WebXRInputSource;
+            squeezeComponent?: WebXRControllerComponent;
+            selectionComponent?: WebXRControllerComponent;
+            onButtonChangedObserver?: Nullable<Observer<WebXRControllerComponent>>;
+            onSqueezeButtonChangedObserver?: Nullable<Observer<WebXRControllerComponent>>;
+            onFrameObserver?: Nullable<Observer<XRFrame>>;
+            meshUnderPointer: Nullable<AbstractMesh>;
+            pick: Nullable<PickingInfo>;
+            id: number;
+            pickIndexMeshTip: Nullable<AbstractMesh>;
+            hoverIndexMeshTip: Nullable<AbstractMesh>;
+            grabRay: Ray;
+            nearInteraction: boolean;
+            nearGrab: boolean;
+            nearGrabInProcess: boolean;
+            // event support
+            eventListeners?: { [event in XREventType]?: (event: XRInputSourceEvent) => void };
+        };
+    } = {};
+    private _scene: Scene;
+
+    private _attachedController: string;

```

Comment:
>_attachedController [](http://example.com/codeflow?start=12&length=19)

There can only be one controller used for near interaction at a time?

Reply:
Explained above. in short no, there can be more than one.